docs/adr/0026-expand-subscription-platform-metadata.md

```diff
@@ -0,0 +1,81 @@
+# Expand Subscription Platform Metadata with Google Cloud Firestore
+
+- Status: proposed
+- Deciders: Ben Bangert, Barry Chen, Bianca Danforth
+- Date: 2021-06-24
+
+## Context and Problem Statement
+
+The Subscription Platform uses [Stripe Metadata][stripe-metadata] for numerous
+purposes. However, Stripe metadata has a limit on the number of values that we
+will reach in the near future. In order to continue to use metadata in the
+same fashion in the Subscription Platform, we need to overcome the limit placed
+by Stripe.
```

Comment:
> 
> 
> > The Subscription Platform uses [Stripe Metadata][stripe-metadata] for numerous
> > purposes.
> 
> Can you list a few purposes here to be more explicit? Who updates this metadata today?

Sure.

> 
> > However, Stripe metadata has a limit on the number of values that we will reach in the near future.
> 
> Maybe we can link to the Stripe docs here that discuss that limit?

Yep.


> How close are we to this limit today? What is the timeline in which we expect to exceed that limit? More broadly, are there other non-idealities to relying on Stripe metadata than this limit (e.g. metadata validation)?
> 

The number of metadata keys used is product and pricing plan dependent.  I could give some kind of napkin sketch math here if you think that'd be helpful.

As for the timeline, that's purely a Product thing; I don't truly know.  Would that affect the decision?

Validation can be a whole separate topic altogether, regardless of storage, transport, access control, etc.  We can discuss more in a future jira (the one I requested Dave to file).  It's not a basis for the decision here is what I think I'm trying to say.

Reply:
> The number of metadata keys used is product and pricing plan dependent. I could give some kind of napkin sketch math here if you think that'd be helpful.
>
> As for the timeline, that's purely a Product thing; I don't truly know. Would that affect the decision?

Hm... no I don't think that'd be helpful here. I'm just curious about these things from an urgency perspective, but you're right that this sort of information is not the most relevant to include here.

> Validation can be a whole separate topic altogether, regardless of storage, transport, access control, etc. We can discuss more in a future jira (the one I requested Dave to file). It's not a basis for the decision here is what I think I'm trying to say.

Why wouldn't validation (before saving) be a basis for a decision here -- particularly if we are planning to go with an approach that does not provide a direct review/approval process? That seems like a "pro" to me, even if it isn't part of an initial implementation... Is it because pretty much all the approaches we have thus far considered have some mechanism to validate the data (unlike the Stripe metadata approach we are using currently)?

Reply:
> Why wouldn't validation (before saving) be a basis for a decision here -- particularly if we are planning to go with an approach that does not provide a direct review/approval process? ...

The validation rules are from us, regardless the form they take, the way they are managed, or the context in which they are executed.  The feature that Firestore provide seems great (at least from some quick reading of their docs); I'll add it as a pro.  But generally, none of the options would prevent us from performing validation... or, what you said, and said it better.

Reply:
> 
> 
> > Maybe we can link to the Stripe docs here that discuss that limit?
> 
> Yep.
> 
I just realized that the doc linked in the ADR discusses that limit.  There isn't another page to link.